packages/editor/src/components/post-author/index.js

```diff
@@ -1,49 +1,165 @@
 /**
  * WordPress dependencies
  */
-import { __ } from '@wordpress/i18n';
+import { sprintf, __, _n } from '@wordpress/i18n';
 import { withInstanceId, compose } from '@wordpress/compose';
 import { Component } from '@wordpress/element';
 import { withSelect, withDispatch } from '@wordpress/data';
+import apiFetch from '@wordpress/api-fetch';
 
 /**
  * Internal dependencies
  */
 import PostAuthorCheck from './check';
 
+/**
+ * External dependencies
+ */
+import Autocomplete from 'accessible-autocomplete/react';
+
+import { debounce } from 'lodash';
+
 export class PostAuthor extends Component {
 	constructor() {
 		super( ...arguments );
 
 		this.setAuthorId = this.setAuthorId.bind( this );
+		this.suggestAuthor = this.suggestAuthor.bind( this );
+		this.getCurrentAuthor = this.getCurrentAuthor.bind( this );
+		this.resolveResults = this.resolveResults.bind( this );
+		this.state = {
+			postAuthor: false,
+		};
+		this.searchCache = [];
+		this.requestResults = debounce( ( query, populateResults ) => {
+			const payload = '?search=' + encodeURIComponent( query );
+			apiFetch( { path: `/wp/v2/users${ payload }` } ).then( ( results ) => {
+				populateResults( this.resolveResults( results ) );
+				this.searchCache[ query ] = results;
+			} );
+		}, 300 );
+		this.requestResults = this.requestResults.bind( this );
+	}
+
+	componentDidMount() {
+		const { postAuthorId, authors, postAuthor } = this.props;
+
+		// If the postAuthor is provided, use it directly.
+		if ( postAuthor ) {
+			this.setState( { postAuthor } );
+		}
+
+		const authorInAuthors = authors.find( ( singleAuthor ) => {
+			return singleAuthor.id === postAuthorId;
+		} );
+
+		// Set or fetch the current author.
+		if ( authorInAuthors ) {
+			this.setState( { postAuthor: authorInAuthors } );
+		} else {
+			this.getCurrentAuthor( postAuthorId );
+		}
+	}
+
+	suggestAuthor( query, populateResults ) {
+		if ( query.length < 2 ) {
+			return;
+		}
+
+		if ( this.searchCache[ query ] ) {
+			populateResults( this.resolveResults( this.searchCache[ query ] ) );
+			return;
+		}
+
+		this.requestResults( query, populateResults );
+	}
+
+	resolveResults( results ) {
+		this.authors = results;
+		return results.map( ( author ) => ( author.name ) );
+	}
+
+	getCurrentAuthor( authorId ) {
+		if ( ! authorId ) {
+			return;
+		}
+		apiFetch( { path: `/wp/v2/users/${ encodeURIComponent( authorId ) }` } ).then( ( results ) => {
+			this.setState( { postAuthor: results } );
+		} );
 	}
 
-	setAuthorId( event ) {
+	setAuthorId( selection ) {
+		if ( ! selection ) {
+			return;
+		}
 		const { onUpdateAuthor } = this.props;
-		const { value } = event.target;
-		onUpdateAuthor( Number( value ) );
+		if ( typeof selection === 'string' ) {
+			// Author name from the autocompleter.
+			const author = this.authors.find( ( singleAuthor ) => {
+				return singleAuthor.name === selection;
+			} );
+			onUpdateAuthor( Number( author.id ) );
```

Comment:
Do we have a guarantee that `author` is not `undefined` here from the result of the prior `Array#find`? Otherwise this will throw an error.

Reply:
Yes, I think we do. This handler gets the selection, which the author name, then searches the list of options (of which the selection is one) to extract the id. one potential issue though would be duplicate names. I'll try to change this to pass the id instead.

Reply:
> I'll try to change this to pass the id instead.

Did this happen?

Reply:
Digging into this a bit, the current implementation does not support passing an id to the searching function - only the value; Tis made me concerned about duplicate user names which led me to reviewing the current ("classic editor") implementation which I re-discovered includes the (unique) user slug as part of the select values:
![](https://cl.ly/2d76ea94afe4/Monosnap_2019-06-25_16-05-44.jpg)

I updated the PR here to match, guaranteeing unique author 'names' - actually "Name (slug)". In addition, I added a check to ensure on a successful author selection would be set.